text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,607 @@
+   necessary. Note that values moved from in step 1 are _not_ dropped.
+3. The caller's stack frame is removed from the stack.
+4. Control is transferred to the callee's entry point.
+
+This implies that it is invalid for any references into the caller's stack frame to outlive the call. The borrow checker ensures that none of the above steps will result in the use of a value that has gone out of scope.
+
+As `become` is always in a tail position (due to being used in place of `return`), this requirement for TCO is already
+fulfilled.
+
+Example:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    become y(a)
+}
+```
+
+Will be desugared in the following way:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    let _tmp = a;
+    drop(b);
+    become y(_tmp)
+}
+```
+
+## Implementation
+[implementation]: #implementation
+
+A now six years old implementation for the earlier mentioned
+[RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md) can be found at
+[DemiMarie/rust/tree/explicit-tailcalls](https://github.com/DemiMarie/rust/tree/explicit-tailcalls).
+A new implementation is planned as part of this RFC.
+
+The parser parses `become` exactly how it parses the `return` keyword. The difference in semantics is handled later.
+
+During type checking, the following are checked:
+
+1. The target of the tail call is, in fact, a simple call.
+2. The target of the tail call has the proper ABI.
+
+Later phases in the compiler assert that these requirements are met.
+
+New nodes are added in HIR and THIR to correspond to `become`. In MIR, the function call is checked that:
+1. The returned value is directly returned.
+2. There are no cleanups.
+3. The basic block being branched into has length zero.
+4. The basic block being branched into terminates with a return.
+
+If these conditions are fulfilled the function call following `become` is flagged to indicate the TCO requirement. This
+flag is then propagated to the corresponding backend. In the backend, there is an additional check if TCO can be
+performed.
+
+Should any check during compilation not pass a compiler error should be issued.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+<!-- Why should we *not* do this? -->
+As this feature should be mostly independent of other features the main drawback lies in the implementation and
+maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
+tooling. The primary effort, however, lies in supporting this feature in the backends:
+- LLVM supports a `musttail` marker to indicate that TCO should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature, seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 30.03.23).
+- GCC does not support an equivalent `musttail` marker.
```

Comment:
I did a bit of research and actually it seems to be supported, but only via the libgccjit API (which rustc_codegen_gcc uses): https://gcc.gnu.org/onlinedocs/gcc-7.3.0/jit/topics/expressions.html#gcc_jit_rvalue_set_bool_require_tail_call

Since this isn't exposed as an attribute to C code I don't have an easy way to test how well it works.

Reply:
Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/ae2f3b80d317a6352442f7a747aa4f34ef0a7f20

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/ae2f3b80d317a6352442f7a747aa4f34ef0a7f20